FUSETOOLS-2169 - adding preference page for staging repositories

bfitzpat opened the pull request.

bfitzpat commented:
@apupier and I have already gone through all the settings. Our theory is that some of the files were checked in with the wrong LF to start with. I am committing things with linux LF now - we ensured that. So I'm not sure how to help you make this better.

bfitzpat commented:
Sorry I didn't realize https://github.com/fusesource/fuseide/pull/730 was out there already. I think the editable list is a little cleaner than a single-line text box, but I don't care which one we go with really.

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/preferences/StagingRepositoryListEditor.java:
```diff
@@ -0,0 +1,105 @@
+	protected String[] parseString(String stringList) {
+		StringTokenizer st = new StringTokenizer(stringList,
+				ProjectTemplatesPreferencesConstants.REPO_SEPARATOR + "\n\r");//$NON-NLS-1$
+		ArrayList<Object> v = new ArrayList<>();
```
ArrayList<Object> --> List<String> seems more precise isn't it?

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/l10n/messages.properties:
```diff
@@ -1,69 +1,78 @@
+newProjectWizardTemplatePageFilterBoxText = type filter text (filters on name, description and keywords)
+UnzipStreamCreator_UnzippingTemplateFileMonitorMessage=Unzipping template files...
+
+stagingRepositoriesPreferencePageDescription=Preferences used when specifying Staging Repositories...
```
i would just put : instead of ... at the end.

I would be more precise on why and where it is used, something like:
"Provide ability to specify extra repositories which will be added to Camel project. This is useful when you are working with staging repositories for instance."

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/OSGI-INF/l10n/bundle.properties:
```diff
@@ -1,36 +1,37 @@
\ No newline at end of file
+template.medium.cxf.code.first.description = This template provides a sample Camel route definition which is started by a CXF Webservice call.
+extension-point.name = projectTemplates
+integration.preferences.project.templates=Project Templates
```
it is nto affecting only Project Templates, it is also used when the facet version is applied/modified.
Using aa name such as "Staging repositories" might be more precise

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/preferences/ProjectTemplatesPreferencePage.java:
```diff
@@ -0,0 +1,108 @@
+	/**
+	 * Boolean field for turning staging repositories on/off.
+	 */
+	private BooleanFieldEditor enableStagingRepositorieslEditor;
```
this attribute is not used, we can use a local variable

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/preferences/ProjectTemplatesPreferencePage.java:
```diff
@@ -0,0 +1,108 @@
+	 * List field for modifying the list of staging repositories and their
+	 * order.
+	 */
+	private StagingRepositoryListEditor stagingRepositoriesListEditor;
```
this attribute is not used, we can use a local variable

bfitzpat commented:
@apupier Can you point me at a good example of the types of tests you'd like to see for this and some examples of how to implement them?

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/preferences/ProjectTemplatesPreferencesConstants.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * @author bfitzpat
+ */
+public interface ProjectTemplatesPreferencesConstants {
```
it is not impacting only project templates, so we might rename it to something like StagingRepositoriesConstants

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/preferences/StagingRepositoryDialog.java:
```diff
@@ -0,0 +1,181 @@
+	private String strName = ""; //$NON-NLS-1$
+
+	// repository url
+	private Text txtURL;
```
What do you think of renaming it txtRepositoryURL instead of providing a comment?

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/preferences/StagingRepositoryDialog.java:
```diff
@@ -0,0 +1,181 @@
+public class StagingRepositoryDialog extends Dialog {
+
+	// repository name
+	private Text txtName;
```
What do you think of renaming it txtRepositoryName instead of providing a comment?

bfitzpat commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/preferences/ProjectTemplatesPreferencePage.java:
```diff
@@ -0,0 +1,108 @@
+		addField(this.stagingRepositoriesListEditor);
+
+		// Sets up the context sensitive help for this page
+		// PlatformUI.getWorkbench().getHelpSystem().setHelp(getFieldEditorParent(), "org.fusesource.ide.camel.editor.editorConfig");
```
Not useless. We will need to hook help up to this preference page and this is a good indicator that remains to be done.

apupier replied:
nobody will notice it. If some tasks are still to do we need to raise another JIRA.

bfitzpat replied:
Ok

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/preferences/StagingRepositoryDialog.java:
```diff
@@ -0,0 +1,181 @@
+				Text textWidget = (Text) e.getSource();
+				String urlText = textWidget.getText();
+				strURL = urlText;
+				getButton(IDialogConstants.OK_ID).setEnabled(isValidURL(strURL));
```
I think that name also need to be not empty (and not only containing spaces)

bfitzpat commented:
@apupier I have addressed all of your suggestions I think except for the testing one. Still not quite sure how you want that handled. You already have a unit test for the preferences themselves, just not for the new page. If you can point me to an example of an integration test that runs a preference page through its paces, I'll happily write some.

bfitzpat commented:
@apupier Made a first attempt at some integration tests for the new preference page. Had to expose an additional package and create a few methods so things were available in the test class, but it's a good first cut I think.

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/preferences/StagingRepositoryDialog.java:
```diff
@@ -0,0 +1,232 @@
+	@Override
+	public void create() {
+		super.create();
+	    setTitle(Messages.NewStagingRepositoryDialogTitle);
```
optional: little indentation issue

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/preferences/StagingRepositoryDialog.java:
```diff
@@ -0,0 +1,232 @@
+	protected Control createDialogArea(Composite parent) {
+		Composite area = (Composite) super.createDialogArea(parent);
+		Composite container = new Composite(area, SWT.NONE);
+	    container.setLayoutData(new GridData(GridData.FILL_BOTH));
```
optional: indentation issue, spaces are used instead of tabs

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/preferences/StagingRepositoryDialog.java:
```diff
@@ -0,0 +1,232 @@
+					setErrorMessage(Messages.NewRepoDialog_nameinvalid);
+				} else {
+					setErrorMessage(null);
+//				    resetMessage();	    
```
remove unused commented code

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/preferences/StagingRepositoryDialog.java:
```diff
@@ -0,0 +1,232 @@
+					setErrorMessage(Messages.NewRepoDialog_urlinvalid);
+				} else {
+					setErrorMessage(null);
+//				    resetMessage();	    
```
remove unused commented code

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/preferences/StagingRepositoryDialog.java:
```diff
@@ -0,0 +1,232 @@
+	}
+
+	@Override
+	protected Point getInitialSize() {
```
as the super is directly called, the whole method can be removed

apupier commented on editor/tests/org.fusesource.ide.projecttemplates.tests.integration/src/main/java/org/fusesource/ide/projecttemplates/tests/preferences/StagingRepositoriesPreferencePageIT.java:
```diff
@@ -0,0 +1,128 @@
+		preferenceStore.setValue(StagingRepositoriesConstants.STAGING_REPOSITORIES, testList);
+
+		// make sure they set properly
+		assertThat(preferenceStore.getBoolean(StagingRepositoriesConstants.ENABLE_STAGING_REPOSITORIES) == true);
```
assertThat(preferenceStore.getBoolean(StagingRepositoriesConstants.ENABLE_STAGING_REPOSITORIES)).isTrue()

SonarLint plugin is reporting this kind of issue

apupier commented on editor/tests/org.fusesource.ide.projecttemplates.tests.integration/src/main/java/org/fusesource/ide/projecttemplates/tests/preferences/StagingRepositoriesPreferencePageIT.java:
```diff
@@ -0,0 +1,128 @@
+		assertThat(prefsDialog).isNotNull();
+		Object page = prefsDialog.getSelectedPage();
+		assertThat(page).isNotNull();
+		assertThat(page instanceof StagingRepositoriesPreferencePage);
```
assertThat(page).isInstanceOf(StagingRepositoriesPreferencePage.class)

apupier commented on editor/tests/org.fusesource.ide.projecttemplates.tests.integration/src/main/java/org/fusesource/ide/projecttemplates/tests/preferences/StagingRepositoriesPreferencePageIT.java:
```diff
@@ -0,0 +1,128 @@
+		// make sure that the checkbox is checked and the list is what we expect it to be
+		assertThat(stagingPage.getEnableStagingRepositoriesEditor().getBooleanValue() == preferenceStore.getBoolean(StagingRepositoriesConstants.ENABLE_STAGING_REPOSITORIES));
+		assertThat(stagingPage.getStagingRepositoryListEditor().getItemList().equals(testList));
+		assertThat(stagingPage.getStagingRepositoryListEditor().isListEnabled());
```
assertThat(stagingPage.getStagingRepositoryListEditor().isListEnabled()).isTrue();

apupier commented on editor/tests/org.fusesource.ide.projecttemplates.tests.integration/src/main/java/org/fusesource/ide/projecttemplates/tests/preferences/StagingRepositoriesPreferencePageIT.java:
```diff
@@ -0,0 +1,128 @@
+		
+		// make sure that the checkbox is checked and the list is what we expect it to be
+		assertThat(stagingPage.getEnableStagingRepositoriesEditor().getBooleanValue() == preferenceStore.getBoolean(StagingRepositoriesConstants.ENABLE_STAGING_REPOSITORIES));
+		assertThat(stagingPage.getStagingRepositoryListEditor().getItemList().equals(testList));
```
assertThat(stagingPage.getStagingRepositoryListEditor().getItemList()).isEqualTo(testList);

apupier commented on editor/tests/org.fusesource.ide.projecttemplates.tests.integration/src/main/java/org/fusesource/ide/projecttemplates/tests/preferences/StagingRepositoriesPreferencePageIT.java:
```diff
@@ -0,0 +1,128 @@
+		
+		// make sure that the list control is disabled
+		assertThat(stagingPage.getEnableStagingRepositoriesEditor().getBooleanValue() == preferenceStore.getBoolean(StagingRepositoriesConstants.ENABLE_STAGING_REPOSITORIES));
+		assertThat(!stagingPage.getStagingRepositoryListEditor().isListEnabled());
```
assertThat(stagingPage.getStagingRepositoryListEditor().isListEnabled()).isFalse();

apupier commented on editor/tests/org.fusesource.ide.projecttemplates.tests.integration/src/main/java/org/fusesource/ide/projecttemplates/tests/preferences/StagingRepositoriesPreferencePageIT.java:
```diff
@@ -0,0 +1,128 @@
+		stagingPage = (StagingRepositoriesPreferencePage) page;
+		
+		// make sure that the list control is disabled
+		assertThat(stagingPage.getEnableStagingRepositoriesEditor().getBooleanValue() == preferenceStore.getBoolean(StagingRepositoriesConstants.ENABLE_STAGING_REPOSITORIES));
```
assertThat(stagingPage.getEnableStagingRepositoriesEditor().getBooleanValue()).isEqualTo(preferenceStore.getBoolean(StagingRepositoriesConstants.ENABLE_STAGING_REPOSITORIES));

apupier commented on editor/tests/org.fusesource.ide.projecttemplates.tests.integration/src/main/java/org/fusesource/ide/projecttemplates/tests/preferences/StagingRepositoriesPreferencePageIT.java:
```diff
@@ -0,0 +1,128 @@
+		
+		// close it
+		boolean isClosed = prefsDialog.close();
+		assertThat(isClosed == true);
```
assertThat(prefsDialog.close()).isTrue();

apupier commented on editor/tests/org.fusesource.ide.projecttemplates.tests.integration/src/main/java/org/fusesource/ide/projecttemplates/tests/preferences/StagingRepositoriesPreferencePageIT.java:
```diff
@@ -0,0 +1,128 @@
+	
+	@Test
+	public void testNewRepositoryDialog() throws Exception {
+		PlatformUI.getWorkbench().getDisplay().asyncExec(new Runnable() {
```
as it is run in async and there is not sync after, the test will be always sucessful

apupier commented on editor/tests/org.fusesource.ide.projecttemplates.tests.integration/src/main/java/org/fusesource/ide/projecttemplates/tests/preferences/StagingRepositoriesPreferencePageIT.java:
```diff
@@ -0,0 +1,128 @@
+				assertThat(dialog.getErrorMessage()).isNull();
+				
+				dialog.pressOK();
+				assertThat(dialog.getName().equals("valid"));
```
assertThat(dialog.getName()).isEqualTo("valid");

bfitzpat commented:
Ok. I've updated the validation so it handles all the cases in a better fashion. And I updated the tests accordingly. I think I've addressed this next wave of comments anyway. :)

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/META-INF/MANIFEST.MF:
```diff
@@ -46,14 +46,17 @@ Require-Bundle: org.eclipse.ui;bundle-version="3.106.0",
  org.fusesource.ide.projecttemplates.adopters.util,
  org.fusesource.ide.projecttemplates.impl.medium,
  org.fusesource.ide.projecttemplates.impl.simple,
+ org.fusesource.ide.projecttemplates.internal,
```
exporting the internal package? if required by some test plugin, it is better to use x-friends to reduce visibility (even if we are not currently very strict everywhere...)

bfitzpat replied:
Updated to x-friends and seems to work great. Thanks for the suggestion.

lhein commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/preferences/StagingRepositoryListEditor.java:
```diff
@@ -0,0 +1,115 @@
+			}
+			lastName = outName;
+			lastURL = outUrl;
+			return outName + StagingRepositoriesConstants.NAME_URL_SEPARATOR + outUrl;
```
I don't like the notation <repositoryname>;<repositoryurl>.
Can we maybe switch to something more eye friendly like <repositoryName> (<repositoryUrl>) ?

apupier replied:
it is to match the serialization, it will require several other changes to use a different notation for the display as currently no Viewer is used and it is reusing the JFace ListEditor

lhein replied:
I would do the change as it makes it more readable.

apupier replied:
I think that it implies to write a full custom preference page then (and throw almost all of this PR) but maybe I'm wrong. Perhaps Brian who worked on it has an idea which doesn't require it.
We need to keep in mind that these parameters are targeting mostly internal users who are using versions not yet released.

lhein replied:
ok, I am fine with it then. Do we hide that preference page for .Final versions then and how is it solved?

lhein replied:
Maybe I commented on the wrong LoC? I find the displayed list content quite ugly. https://cloud.githubusercontent.com/assets/1105127/20298559/eae6f57a-ab16-11e6-81e6-585dec8dc734.png
Thats what I was basically talking about :D

bfitzpat replied:
The example you're pointing to was due to some validation not being in place yet. It's a little better now. Just a comma between the name and URL shows up as a list item.

![pasted image 1](https://cloud.githubusercontent.com/assets/530878/20353956/cd846dbe-abd9-11e6-9973-5db6d727a1d0.png)

We may be able to do more with the list with a label provider somehow, but I don't think it's that awful. Could it be better? Sure. But not sure it's a priority tbh.

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/preferences/StagingRepositoryListEditor.java:
```diff
@@ -0,0 +1,115 @@
+				StagingRepositoriesConstants.REPO_SEPARATOR + "\n\r");//$NON-NLS-1$
+		
+		List<String> v = new ArrayList<String>();
+		while (st.hasMoreElements()) {
```
Optional: I think that we can use st.HasMoreTokens() and st.nextToken() to avoid cast to String

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/preferences/StagingRepositoryListEditor.java:
```diff
@@ -0,0 +1,115 @@
+		if (rtn == Window.OK && dialog.getURL() != null && dialog.getName() != null) {
+			String outUrl = dialog.getURL().trim();
+			String outName = dialog.getName().trim();
+			if (outUrl.length() == 0 || outName.length() == 0) {
```
optional: I personally prefer the notation outUrl.isEmpty() || outName.isEmpty()

bfitzpat commented:
I'm updating the PR now, but this is where we're at. Without some major rejiggering I don't know how we will be able to change the dialog to be exactly what they're after in the Patternfly example. We might be able to use some styling using CSS (see FUSETOOLS-2191) but I suspect there are ripples there to be aware of.

This document has a bunch of screen shots showing the various extra labels for warnings and such. 

[Staging Repositories preference page.pdf](https://github.com/fusesource/fuseide/files/595612/Staging.Repositories.preference.page.pdf)

Still working on "prettying" up the list so that it shows things a bit more cleanly

bfitzpat commented:
Without completely gutting StagingRepositoryListEditor and rolling our own, I don't see a way to make that list any prettier. It's just a string list essentially. If it's really that important, we can roll our own with a viewer instead of the List control, just so we can add a Label provider on top of it. 

But I think I've addressed all the rest of the new wave of comments.

lhein commented:
I am still wondering what happens once we go GA. Do we hide that properties page then?

apupier commented:
No, i think we should not hide this page for GA.
There will be few days (weeks?) before artifacts will be published on
public repository and Final bits will be provided.

On Thu, Nov 17, 2016 at 7:49 AM, Lars Heinemann <notifications@github.com>
wrote:

> I am still wondering what happens once we go GA. Do we hide that
> properties page then?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fusesource/fuseide/pull/758#issuecomment-261169424>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABDc56F_WI3xLBSmQRVGCZxb9FZgfJbFks5q-_jugaJpZM4KvSSl>
> .
>



-- 
Aurelien Pupier
Senior Software Engineer in JBoss Fuse Tooling team
@apupier

lhein commented:
How about restricting the visibility of that preference page by analyzing the version of the bundle? If it ends with .Final we hide it, otherwise we show it?

apupier commented:
we will be exactly in the case that I describe, few days of blackout after
Final bits provided but repository not publicly published.

I would prefer to keep this page anyway in case some artifacts are missing
from public repository.

On Thu, Nov 17, 2016 at 9:29 AM, Lars Heinemann <notifications@github.com>
wrote:

> How about restricting the visibility of that preference page by analyzing
> the version of the bundle? If it ends with .Final we hide it, otherwise we
> show it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fusesource/fuseide/pull/758#issuecomment-261185426>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABDc5-bOrN-rzt9GYddy6-x7XCCQYLV1ks5q_BBxgaJpZM4KvSSl>
> .
>



-- 
Aurelien Pupier
Senior Software Engineer in JBoss Fuse Tooling team
@apupier

apupier commented on editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/preferences/StagingRepositoryListEditor.java:
```diff
@@ -0,0 +1,98 @@
+
+	@Override
+	protected String[] parseString(String stringList) {
+		StringTokenizer st = new StringTokenizer(stringList,
```
For Another Review: in Stringtokenizer javadoc

> StringTokenizer is a legacy class that is retained for compatibility reasons although its use is discouraged in new code. It is recommended that anyone seeking this functionality use the split method of String or the java.util.regex package instead. 

I have no problem to use it here as the class is not even marked as deprecated. But next time we have a split to do we should try to have a look to some String.split or Streams.

bfitzpat replied:
Ok
